### PR TITLE
Improved anchor handling.

### PR DESCRIPTION
**Pull request by judax**
Verified that the anchor calculations are correct. Found that the main reason why some error were arising is because it is wrong to assume that an anchor can be created with a model matrix that includes scaling that is not (1, 1, 1). Added some new functionalities and more comments but the anchor handling stays the same. The JS side will have to make sure that the anchors are correctly created.

Change-Id: I98de8dadb2c7f0965633eed3c4b3603cf76748c9

### CODE REVIEW

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 1]
```diff
@@ -18,31 +18,44 @@ std::shared_ptr<Anchor> AnchorManager::addAnchor(double timestamp,
...
+  anchorsMutex.unlock();
   return anchor;
 }
 
 void AnchorManager::removeAnchor(uint32_t identifier) {
+  anchorsMutex.lock();
   std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::iterator it = 
       anchors.find(identifier);
   if (it != anchors.end())
   {
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: { should be on previous line

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 2]
```diff
@@ -18,31 +18,44 @@ std::shared_ptr<Anchor> AnchorManager::addAnchor(double timestamp,
...
                                             anchorModelMatrix));
+  anchorsMutex.lock();
   anchors[anchor->getIdentifier()] = anchor;
+  anchorsMutex.unlock();
   return anchor;
 }
 
 void AnchorManager::removeAnchor(uint32_t identifier) {
+  anchorsMutex.lock();
   std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::iterator it =
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: this is a good time to use auto. Alternatively (and my preference) is just say:
if (anchors.find(identifier) != anchors.end()) {
  ...

File: android_webview/test/shell/tango/jni/AnchorManager.cpp [hunk 3]
```diff
@@ -18,31 +18,44 @@ std::shared_ptr<Anchor> AnchorManager::addAnchor(double timestamp,
...
 std::vector<std::shared_ptr<Anchor>> AnchorManager::update(
     double historyChangeTimestamp, int activityOrientation) {
+  // Create a container to hold up to all the anchors that are currently
+  // available. Remember, the capacity is not the size of the container.
   std::vector<std::shared_ptr<Anchor>> updatedAnchors;
   updatedAnchors.reserve(anchors.size());
+  // Iterate over all the anchors and try to update them. Each anchor's
+  // timestamp will indicate if the anchor needs to be updated.
+  anchorsMutex.lock();
   std::unordered_map<uint32_t, std::shared_ptr<Anchor>>::const_iterator it =
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/AnchorManager.cpp]: auto

File: android_webview/test/shell/tango/jni/MathUtils.cpp [hunk 4]
```diff
@@ -2,16 +2,17 @@
...
 namespace tango_chromium {
 
 void matrixFrustum(float const & left,
-             float const & right,
-             float const & bottom,
-             float const & top,
-             float const & near,
-             float const & far,
-             float* matrix) {
+    float const & right,
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/MathUtils.cpp]: isn't it standard to not have space between const and &?

File: android_webview/test/shell/tango/jni/MathUtils.cpp [hunk 5]
```diff
@@ -103,7 +104,87 @@ glm::mat4 mat4FromTranslationOrientation(const double *translation,
   return m;
 };
 
-float rayIntersectsPlane(const glm::vec3 &planeNormal, const glm::vec3 &planePosition,
+void translationOrientationAndScaleFromMat4(const glm::mat4& matrix, 
+    glm::vec3& translation, glm::quat& orientation, glm::vec3& scale) {
+
+  const float* pMatrix = glm::value_ptr(matrix);
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/MathUtils.cpp]: this should be p_matrix or matrix_values or something.

All the variable names in this function (and all our c++ moving forward) should use underscore_case, not camelCase

File: android_webview/test/shell/tango/jni/MathUtils.h [hunk 6]
```diff
@@ -6,29 +6,38 @@
 namespace tango_chromium {
 
 void matrixFrustum(float const & left, float const & right,
-                   float const & bottom, float const & top,
-                   float const & near, float const & far,
-                   float* matrix);
+    float const & bottom, float const & top,
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/MathUtils.h]: const&  ?

File: android_webview/test/shell/tango/jni/MathUtils.h [hunk 7]
```diff
@@ -6,29 +6,38 @@
...
-bool isPointInPolygon(const glm::vec3 &point, const glm::vec3* polygonPoints, int count);
+bool isPointInPolygon(const glm::vec3 &point, const glm::vec3* polygonPoints, 
+    int count);
+
+glm::vec3 getSafeReciprocalScale(const glm::vec3& scale, float tolerance);
+
+glm::mat4 getMatrixRelativeToMatrix(const glm::mat4& originalMatrix, 
+    const glm::mat4&relativeToMatrix, bool useOriginalScale);
 
 } // tango_chromium
```
Comment (lincolnfrog) [inline android_webview/test/shell/tango/jni/MathUtils.h]: thanks for cleaning this file up!

**Review by lincolnfrog — APPROVED** [review]

File: android_webview/test/shell/tango/jni/MathUtils.cpp (hunk 4 above)
Comment (judax) [reply android_webview/test/shell/tango/jni/MathUtils.cpp]: Have to admit that copied and pasted this function signature. I will fix it.

File: android_webview/test/shell/tango/jni/MathUtils.cpp (hunk 5 above)
Comment (judax) [reply android_webview/test/shell/tango/jni/MathUtils.cpp]: Will change it. Thanks.

**Pull request merged**